### PR TITLE
[FLINK-2209] Document linking with jars not in the binary dist

### PR DESCRIPTION
**Pull request by mbalassi**

### CODE REVIEW

File: docs/apis/cluster_execution.md [hunk 1]
```diff
@@ -80,67 +80,73 @@ Note that the program contains custom user code and hence requires a JAR file wi
...
-public static void main(String[] args) throws Exception {
-    ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
+1. The maven assembly plugin builds a so called fat jar cointaining all your dependencies.
+Easy to configure, but is an overkill in many cases. See 
+[usage](http://maven.apache.org/plugins/maven-assembly-plugin/usage.html).
+2. The maven unpack plugin, for unpacking the relevant parts of the dependencies and
+then package it with your code.
 
-    DataSet<String> data = env.readTextFile("hdfs://path/to/file");
+To the the latter for example for the streaming Kafka connector, `flink-connector-kafka`
```
Comment (rmetzger) [inline docs/apis/cluster_execution.md]: the the

File: docs/apis/cluster_execution.md [hunk 2]
```diff
@@ -80,67 +80,73 @@ Note that the program contains custom user code and hence requires a JAR file wi
...
-~~~
-
-### Example
-
-The following illustrates the use of the `RemoteExecutor` with the Scala API:
-
-~~~scala
-def main(args: Array[String]) {
-    val input = TextFile("hdfs://path/to/file")
+1. Either copy the required jar files to the `lib` folder onto all of your TaskManagers.
```
Comment (tillrohrmann) [inline docs/apis/cluster_execution.md]: Maybe add that the TMs have to be restarted.

File: docs/libs/gelly_guide.md [hunk 3]
```diff
@@ -43,6 +43,8 @@ Add the following dependency to your `pom.xml` to use Gelly.
 </dependency>
 ~~~
 
+Note that the Gelly is currently not part of the binary distribution. See linking with it for cluster execution [here](../apis/cluster_execution.html#linking-with-modules-not-contained-in-the-binary-distribution).
```
Comment (tillrohrmann) [inline docs/libs/gelly_guide.md]: no the

File: docs/libs/ml/index.md [hunk 4]
```diff
@@ -69,6 +69,8 @@ Next, you have to add the FlinkML dependency to the `pom.xml` of your project.
 </dependency>
 {% endhighlight %}
 
+Note that the FlinkML is currently not part of the binary distribution. See linking with it for cluster execution [here](../apis/cluster_execution.html#linking-with-modules-not-contained-in-the-binary-distribution).
```
Comment (tillrohrmann) [inline docs/libs/ml/index.md]: no the

File: docs/apis/cluster_execution.md [hunk 5]
```diff
@@ -80,67 +80,73 @@ Note that the program contains custom user code and hence requires a JAR file wi
...
-## Remote Executor
+## Linking with modules not contained in the binary distribution
 
-Similar to the RemoteEnvironment, the RemoteExecutor lets you execute
-Flink programs on a cluster directly. The remote executor accepts a
-*Plan* object, which describes the program as a single executable unit.
+The binary distribution contains jar packages in the `lib` folder that are automatically
+provided to the classpath of your distrbuted programs. Almost all of Flink classes are
+located there with a few exceptions, for example the streaming connectors and some freshly
+added modules. To run code depending on these modules you need to make them accessible
```
Comment (tillrohrmann) [inline docs/apis/cluster_execution.md]: Why listing the streaming connectors and not the other libraries? Maybe we can list/link all of them here.

Comment (mbalassi) [inline docs/apis/cluster_execution.md]: The intention here was to write a sentence that does not have to be updated. We might want to add FlinkML to the binary later, and are introducing new modules frequently, e.g. Storm compatibility will be in the code shortly. I do not think that it is realistic to maintain a sentence here.

File: docs/apis/cluster_execution.md (hunk 2 above)
Comment (mbalassi) [inline docs/apis/cluster_execution.md]: Sure.

File: docs/apis/cluster_execution.md [hunk 6]
```diff
@@ -80,67 +80,73 @@ Note that the program contains custom user code and hence requires a JAR file wi
...
 
-The following illustrates the use of the `RemoteExecutor` with the Java API (as
-an alternative to the RemoteEnvironment):
+To provide these dependencies not included by Flink we suggest two options with Maven.
 
-~~~java
-public static void main(String[] args) throws Exception {
-    ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
+1. The maven assembly plugin builds a so called fat jar cointaining all your dependencies.
+Easy to configure, but is an overkill in many cases. See
```
Comment (tillrohrmann) [inline docs/apis/cluster_execution.md]: There is no subject in the sentence.

Comment (tillrohrmann) [inline docs/apis/cluster_execution.md]: overkill is maybe a little bit too colloquial here

File: docs/apis/cluster_execution.md (hunk 1 above)
Comment (tillrohrmann) [inline docs/apis/cluster_execution.md]: Wording of the first sentence. Maybe something like: "Using the latter approach in order to bundle the Kafka connecter..."

**Comment by mbalassi** [comment]
Thanks for the comments, I have addressed them. If no other objections I would like to merge this in the evening.

**Comment by StephanEwen** [comment]
I am a bit confused here. With the quickstart maven projects, fat jar building works without configuring, as it defines proper assembly and share configurations for fat jars.

This description reads as if one needs to always use the maven unpack plugin when using any dependency not in `lib`.

**Comment by mbalassi** [comment]
I do not think that we can expect that every user starts from the quickstart poms, but it is true that a reference to that would have been nicer. 

I have to admit that I had the old version of quickstart poms in mind when writing this commit. That  excluded anything in the `org.apache.flink` namespace and was obviously not a good solution. Thanks for pointing this out, we should modify this guide then.

**Comment by rmetzger** [comment]
Can you take care of this @mbalassi ?

**Pull request closed without merging**